### PR TITLE
Setting snapshot removed on timeout

### PR DESCRIPTION
**Pull request by Spaceman1984**
## Description

When creating a snapshot times-out on the management server, the record in snapshot_store_ref table is deleted.

By setting the snapshot to "removed" on timeout, the record is not deleted.

Fixes: #3559 
## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

* Set the global variable: backup.snapshot.wait to 2 seconds.
* Create a volume snapshot.
* Check the snapshots table for a snapshot record and check for a corresponding record in snapshot_store_ref table.

### CODE REVIEW

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM

**Review by davidjumani — APPROVED** [review]
LGTM

**Review by davidjumani — COMMENTED** [review]
Discussed an issue with @Spaceman1984 
Might be unrelated, Waiting for his response

**Comment by DaanHoogland** [comment]
> Discussed an issue with @Spaceman1984
> Might be unrelated, Waiting for his response

so this is on hold, @davidjumani ?

**Comment by Spaceman1984** [comment]
I'll convert to draft @DaanHoogland

**Comment by Spaceman1984** [comment]
Snapshots are being deleted from the database when using the "Async" flag on creation and it times out. - This is happening in a separate process - I suspect the snapshots are being marked for garbage collection in the async process. I'll need to look into this a bit further.

**Comment by rhtyd** [comment]
@Spaceman1984 can you share the latest update on this, is this ready for review/testing - or you still investigating?

**Comment by Spaceman1984** [comment]
Still investigating @rhtyd

**Comment by rhtyd** [comment]
@Spaceman1984 any update on this?

**Comment by Spaceman1984** [comment]
No progress yet, I have been pre-occupied with some other work stuff, I was only able to start looking at this again today.

**Comment by Spaceman1984** [comment]
In the storage garbage collection process snapshots in an error state are deleted where the "removed" field is null.

[StorageManagerImpl.java]
```
// remove snapshots in Error state
List<SnapshotVO> snapshots = _snapshotDao.listAllByStatus(Snapshot.State.Error);
for (SnapshotVO snapshotVO : snapshots) {
    try {
        List<SnapshotDataStoreVO> storeRefs = _snapshotStoreDao.findBySnapshotId(snapshotVO.getId());
        for (SnapshotDataStoreVO ref : storeRefs) {
            _snapshotStoreDao.expunge(ref.getId());
        }
        _snapshotDao.expunge(snapshotVO.getId());
    } catch (Exception e) {
        s_logger.warn("Unable to destroy snapshot " + snapshotVO.getUuid(), e);
    }
}
```

This line only looks at where records are not removed
`List<SnapshotVO> snapshots = _snapshotDao.listAllByStatus(Snapshot.State.Error);`


I can do a few things here, I can set the removed field to the current timestamp on error which should prevent the field from being deleted or I can create a different TimedOut state so these records are also not cleaned out by the garbage collection process.

**Comment by rhtyd** [comment]
@Spaceman1984 please update if this is ready for review?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:571]: should this be in a transaction to be sure?

**Review by DaanHoogland — COMMENTED** [review]
code looks good, it would be just one of many fixes i imagine and also; is AncientDataMotionStrategy the only place to do this? aren't there vmware or kvm specific places as well?

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:571]: ping @Spaceman1984

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:571]: ping @davidjumani @GabrielBrascher @rhtyd the two statements above should be grouped in a transaction. agree?

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java
Comment (sureshanaparti) [inline engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:593]: @Spaceman1984 the copy callback should mark the snapshot state accordingly, based on the operation result (success / fail). Not required to set the state explicitly.

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:568]: @Spaceman1984 can you handle this in the copy callback, and pass the proper event to set the snapshot state?

Comment (Spaceman1984) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:568]: Not sure what you mean @sureshanaparti

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:568]: > Not sure what you mean @sureshanaparti

@Spaceman1984 the caller of _copyAsync_ method would set a callback method to process the result and set the object state accordingly. Can you use that callback method (of snapshot copy operation) to set the state, through the event (check the snapshot state machine) ?

Comment (Spaceman1984) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:568]: I'll take a look.

**Comment by Spaceman1984** [comment]
> code looks good, it would be just one of many fixes i imagine and also; is AncientDataMotionStrategy the only place to do this? aren't there vmware or kvm specific places as well?

I tested this specifically with a vmware environment as the original issue was reported on vmware @DaanHoogland.

**Comment by Spaceman1984** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java
Comment (Spaceman1984) [reply server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java]: I have removed setting the error state, this leaves the snapshot in a "backing up" state forever which is fine.

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java
Comment (Spaceman1984) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java]: Addressed

**Comment by Spaceman1984** [comment]
@blueorangutan test help

**Comment by Spaceman1984** [comment]
@blueorangutan test matrix

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM

**Review by DaanHoogland — APPROVED** [review]
code looks good. not tested

**Comment by DaanHoogland** [comment]
@borisstoyanov this pertains to your #3559. Can you give the final stamp of approval?

**Comment by DaanHoogland** [comment]
@sureshanaparti please validate you concerns are indeed met.

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:556]: _markRemovedSnapshot_ indicates the snapshot is already removed. I think _removeSnapshot_ (or _markSnapshotAsRemoved_) is ok.

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:566]: move caller here.

```suggestion
        Account caller = CallContext.current().getCallingAccount();
        _accountMgr.checkAccess(caller, null, true, snapshotCheck);
```

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java
Comment (sureshanaparti) [inline engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:593]: Can you check for OperationTimedoutException (if thrown for timeout?), set removed accordingly. Note that, the snapshot here is not in Error state as the copy callback is not yet to be called, which takes the action based on the operation failure.

Comment (Spaceman1984) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:593]: I think having the snapshot marked as removed here in the case of any exception is better than just checking for a timeout. I think if the process fails for any reason we would want to keep the db records.

**Comment by rhtyd** [comment]
@blueorangutan package

@sureshanaparti @borisstoyanov any final comments

**Comment by Spaceman1984** [comment]
@blueorangutan test matrix

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java
Comment (sureshanaparti) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java]: @Spaceman1984 possible to move this method call to copy snapshot callback method on result failure?

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java
Comment (Spaceman1984) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java]: Thanks @sureshanaparti for pointing me in that direction, I was able to get rid of that method.

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test matrix

**Review by sureshanaparti — APPROVED** [review]
changes LGTM

**Comment by DaanHoogland** [comment]
prety sure the kvm failures are environmental, rerunning that one
@blueorangutan test centos7 kvm-centos7

**Review by davidjumani — APPROVED** [review]
LGTM. Snaps removed when operation timed out when async set to both true and false

**Comment by rhtyd** [comment]
@DaanHoogland let's merge this once the test results are back we've enough LGTMs and manual tests done

**Pull request merged**
